src/platform/audio/ffmpegaudiofilecache.cpp

```diff
@@ -0,0 +1,537 @@
+            return nullptr;
+        default:
+            break;
+    }
+
+    captainslog_trace("FFmpegAudioFileCache: opening file %s", filename.Str());
+
+    // Try to find existing data for this file to avoid loading it if unneeded.
+    auto it = m_cacheMap.find(filename);
+
+    if (it != m_cacheMap.end()) {
+        ++it->second.ref_count;
+
+        return it->second.wave_data;
+    }
+
+    // Load the file from disk
+    File *file = g_theFileSystem->Open(filename, File::READ | File::BINARY | File::BUFFERED);
+
+    if (file == nullptr) {
+        if (!filename.Is_Empty()) {
+            captainslog_warn("Missing audio file '%s', could not cache.", filename.Str());
+        }
+
+        return nullptr;
+    }
+
+    FFmpegOpenAudioFile open_audio;
+    open_audio.wave_data = static_cast<uint8_t *>(av_malloc(sizeof(WavHeader)));
+    open_audio.data_size = sizeof(WavHeader);
+    open_audio.audio_event_info = audio_event->Get_Event_Info();
+
+    if (!Open_FFmpeg_Contexts(&open_audio, file)) {
+        captainslog_warn("Failed to load audio file '%s', could not cache.", filename.Str());
+        return nullptr;
+    }
+
+    if (audio_event->Is_Positional_Audio() && open_audio.codec_ctx->channels > 1) {
+        captainslog_error("Audio marked as positional audio cannot have more than one channel.");
+        return nullptr;
+    }
+
+    if (!Decode_FFmpeg(&open_audio)) {
+        captainslog_warn("Failed to decode audio file '%s', could not cache.", filename.Str());
+        Close_FFmpeg_Contexts(&open_audio);
+        return nullptr;
+    }
+
+    Fill_Wave_Data(&open_audio);
+    Close_FFmpeg_Contexts(&open_audio);
+
+    open_audio.ref_count = 1;
+    m_currentSize += open_audio.data_size;
+
+    // m_maxSize prevents using overly large amounts of memory, so if we are over it, unload some other samples.
+    if (m_currentSize > m_maxSize && !Free_Space_For_Sample(open_audio)) {
+        captainslog_warn("Cannot play audio file since cache is full: %s", filename.Str());
+        m_currentSize -= open_audio.data_size;
+        Release_Open_Audio(&open_audio);
+
+        return nullptr;
+    }
+
+    m_cacheMap[filename] = open_audio;
+
+    return open_audio.wave_data;
+}
+
+/**
+ * Closes a file, reducing the references to it. Does not actually free the cache.
+ */
+void FFmpegAudioFileCache::Close_File(void *file)
+{
+    if (file == nullptr) {
+        return;
+    }
+
+    ScopedMutexClass lock(&m_mutex);
+
+    for (auto it = m_cacheMap.begin(); it != m_cacheMap.end(); ++it) {
+        if (it->second.wave_data == file) {
+            --it->second.ref_count;
+
+            break;
+        }
+    }
+}
+
+/**
+ * Sets the maximum amount of memory in bytes that the cache should use.
+ */
+void FFmpegAudioFileCache::Set_Max_Size(unsigned size)
+{
+    ScopedMutexClass lock(&m_mutex);
+    m_maxSize = size;
+}
+
+/**
+ * Attempts to free space by releasing files with no references
+ */
+unsigned FFmpegAudioFileCache::Free_Space()
```

Comment:
All the other functions have a mutex should there be one for this (and the other free function)?

Reply:
The free functions are called from functions that already lock themselves - This free function is currently only used in tests

Reply:
Where is this function called? I can't find it here with Browser Text Search. Also, this function is `public`. Shouldn't it be private if it is meant to be called internally only?

Reply:
It's called by the audiofilecache tests. See `test_audiofilecache.cpp`